SAST Report on member function ‘bool CIccFormulaCurveSegmentXml::ToXml(std::string&, std::string)’

## Maintainer Report

Mon Jan  5 12:10:43 AM UTC 2026

Issue: SAST Report on member function ‘bool CIccFormulaCurveSegmentXml::ToXml(std::string&, std::string)’
Code: [IccXML/IccLibXML/IccMpeXml.cpp:168:11](https://github.com/InternationalColorConsortium/iccDEV/blame/master/IccXML/IccLibXML/IccMpeXml.cpp#L170)
Commit: https://github.com/InternationalColorConsortium/iccDEV/commit/024b7afd0a8ad1b469a53c8ce149b93b44cdee3c
Log: https://github.com/InternationalColorConsortium/iccDEV/actions/runs/20701096919/job/59423442857#step:7:508

## SAST Log

```
 /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp: In member function ‘bool CIccFormulaCurveSegmentXml::ToXml(std::string&, std::string)’:
  /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:168:52: warning: ‘%s’ directive output may be truncated writing up to 255 bytes into a region of size 233 [-Wformat-truncation=]
    168 |   snprintf(line, bufSize, "<FormulaSegment Start=\"%s\"", icSegPos(buf, bufSize, m_startPoint));
        |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  In file included from /usr/include/stdio.h:980,
                   from /home/runner/work/iccDEV/iccDEV/Build/Cmake/../../IccProfLib/IccIO.h:71,
                   from /home/runner/work/iccDEV/iccDEV/Build/Cmake/../../IccProfLib/IccTag.h:80,
                   from /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccTagXml.h:64,
                   from /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:64:
  In function ‘int snprintf(char*, size_t, const char*, ...)’,
      inlined from ‘bool CIccFormulaCurveSegmentXml::ToXml(std::string&, std::string)’ at /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:168:11:
  /usr/include/x86_64-linux-gnu/bits/stdio2.h:54:35: note: ‘__builtin___snprintf_chk’ output between 25 and 280 bytes into a destination of size 256
     54 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
        |          ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     55 |                                    __glibc_objsize (__s), __fmt,
        |                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     56 |                                    __va_arg_pack ());
        |                                    ~~~~~~~~~~~~~~~~~
  /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp: In member function ‘bool CIccFormulaCurveSegmentXml::ToXml(std::string&, std::string)’:
  /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:171:35: warning: ‘%s’ directive output may be truncated writing up to 255 bytes into a region of size 250 [-Wformat-truncation=]
    171 |   snprintf(line, bufSize, " End=\"%s\"",icSegPos(buf, bufSize, m_endPoint));
        |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  In function ‘int snprintf(char*, size_t, const char*, ...)’,
      inlined from ‘bool CIccFormulaCurveSegmentXml::ToXml(std::string&, std::string)’ at /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:171:11:
  /usr/include/x86_64-linux-gnu/bits/stdio2.h:54:35: note: ‘__builtin___snprintf_chk’ output between 8 and 263 bytes into a destination of size 256
     54 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
        |          ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     55 |                                    __glibc_objsize (__s), __fmt,
        |                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     56 |                                    __va_arg_pack ());
        |                                    ~~~~~~~~~~~~~~~~~
  /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp: In member function ‘bool CIccSampledCurveSegmentXml::ToXml(std::string&, std::string)’:
  /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:275:52: warning: ‘%s’ directive output may be truncated writing up to 255 bytes into a region of size 233 [-Wformat-truncation=]
    275 |   snprintf(line, bufSize, "<SampledSegment Start=\"%s\"", icSegPos(buf, bufSize, m_startPoint));
        |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  In function ‘int snprintf(char*, size_t, const char*, ...)’,
      inlined from ‘bool CIccSampledCurveSegmentXml::ToXml(std::string&, std::string)’ at /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:275:11:
  /usr/include/x86_64-linux-gnu/bits/stdio2.h:54:35: note: ‘__builtin___snprintf_chk’ output between 25 and 280 bytes into a destination of size 256
     54 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
        |          ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     55 |                                    __glibc_objsize (__s), __fmt,
        |                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     56 |                                    __va_arg_pack ());
        |                                    ~~~~~~~~~~~~~~~~~
  /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp: In member function ‘bool CIccSampledCurveSegmentXml::ToXml(std::string&, std::string)’:
  /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:278:35: warning: ‘%s’ directive output may be truncated writing up to 255 bytes into a region of size 250 [-Wformat-truncation=]
    278 |   snprintf(line, bufSize, " End=\"%s\">\n",icSegPos(buf, bufSize, m_endPoint));
        |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  In function ‘int snprintf(char*, size_t, const char*, ...)’,
      inlined from ‘bool CIccSampledCurveSegmentXml::ToXml(std::string&, std::string)’ at /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:278:11:
  /usr/include/x86_64-linux-gnu/bits/stdio2.h:54:35: note: ‘__builtin___snprintf_chk’ output between 10 and 265 bytes into a destination of size 256
     54 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
        |          ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     55 |                                    __glibc_objsize (__s), __fmt,
        |                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     56 |                                    __va_arg_pack ());
        |                                    ~~~~~~~~~~~~~~~~~
  /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp: In member function ‘bool CIccFormulaCurveSegmentXml::ToXml(std::string&, std::string)’:
  /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:168:52: warning: ‘%s’ directive output may be truncated writing up to 255 bytes into a region of size 233 [-Wformat-truncation=]
    168 |   snprintf(line, bufSize, "<FormulaSegment Start=\"%s\"", icSegPos(buf, bufSize, m_startPoint));
        |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  In file included from /usr/include/stdio.h:980,
                   from /home/runner/work/iccDEV/iccDEV/Build/Cmake/../../IccProfLib/IccIO.h:71,
                   from /home/runner/work/iccDEV/iccDEV/Build/Cmake/../../IccProfLib/IccTag.h:80,
                   from /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccTagXml.h:64,
                   from /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:64:
  In function ‘int snprintf(char*, size_t, const char*, ...)’,
      inlined from ‘bool CIccFormulaCurveSegmentXml::ToXml(std::string&, std::string)’ at /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:168:11:
  /usr/include/x86_64-linux-gnu/bits/stdio2.h:54:35: note: ‘__builtin___snprintf_chk’ output between 25 and 280 bytes into a destination of size 256
     54 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
        |          ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     55 |                                    __glibc_objsize (__s), __fmt,
        |                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     56 |                                    __va_arg_pack ());
        |                                    ~~~~~~~~~~~~~~~~~
  /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp: In member function ‘bool CIccFormulaCurveSegmentXml::ToXml(std::string&, std::string)’:
  /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:171:35: warning: ‘%s’ directive output may be truncated writing up to 255 bytes into a region of size 250 [-Wformat-truncation=]
    171 |   snprintf(line, bufSize, " End=\"%s\"",icSegPos(buf, bufSize, m_endPoint));
        |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  In function ‘int snprintf(char*, size_t, const char*, ...)’,
      inlined from ‘bool CIccFormulaCurveSegmentXml::ToXml(std::string&, std::string)’ at /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:171:11:
  /usr/include/x86_64-linux-gnu/bits/stdio2.h:54:35: note: ‘__builtin___snprintf_chk’ output between 8 and 263 bytes into a destination of size 256
     54 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
        |          ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     55 |                                    __glibc_objsize (__s), __fmt,
        |                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     56 |                                    __va_arg_pack ());
        |                                    ~~~~~~~~~~~~~~~~~
  /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp: In member function ‘bool CIccSampledCurveSegmentXml::ToXml(std::string&, std::string)’:
  /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:275:52: warning: ‘%s’ directive output may be truncated writing up to 255 bytes into a region of size 233 [-Wformat-truncation=]
    275 |   snprintf(line, bufSize, "<SampledSegment Start=\"%s\"", icSegPos(buf, bufSize, m_startPoint));
        |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  In function ‘int snprintf(char*, size_t, const char*, ...)’,
      inlined from ‘bool CIccSampledCurveSegmentXml::ToXml(std::string&, std::string)’ at /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:275:11:
  /usr/include/x86_64-linux-gnu/bits/stdio2.h:54:35: note: ‘__builtin___snprintf_chk’ output between 25 and 280 bytes into a destination of size 256
     54 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
        |          ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     55 |                                    __glibc_objsize (__s), __fmt,
        |                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     56 |                                    __va_arg_pack ());
        |                                    ~~~~~~~~~~~~~~~~~
  /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp: In member function ‘bool CIccSampledCurveSegmentXml::ToXml(std::string&, std::string)’:
  /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:278:35: warning: ‘%s’ directive output may be truncated writing up to 255 bytes into a region of size 250 [-Wformat-truncation=]
    278 |   snprintf(line, bufSize, " End=\"%s\">\n",icSegPos(buf, bufSize, m_endPoint));
        |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  In function ‘int snprintf(char*, size_t, const char*, ...)’,
      inlined from ‘bool CIccSampledCurveSegmentXml::ToXml(std::string&, std::string)’ at /home/runner/work/iccDEV/iccDEV/IccXML/IccLibXML/IccMpeXml.cpp:278:11:
  /usr/include/x86_64-linux-gnu/bits/stdio2.h:54:35: note: ‘__builtin___snprintf_chk’ output between 10 and 265 bytes into a destination of size 256
     54 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
        |          ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     55 |                                    __glibc_objsize (__s), __fmt,
        |                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     56 |                                    __va_arg_pack ());
        |                                    ~~~~~~~~~~~~~~~~~
```

## Request

Please consider refactoring and/or adding comments to the code if this is the intended behavior when time permits. If this is the intended behavior please Label as, OOS, wontfix and Close.

Thank You

--- a/IccXML/IccLibXML/IccMpeXml.cpp
+++ b/IccXML/IccLibXML/IccMpeXml.cpp
@@ -85,11 +85,13 @@ bool CIccMpeXmlUnknown::ToXml(std::string &xml, std::string blanks/* = ""*/)
   //icUInt8Number *m_ptr = m_pData;
 
   const size_t bufSize = 256;
-  char line[bufSize];
-  char buf[bufSize];
+  const size_t smallBufSize = 20;
+  char line[bufSize*2];
+  char buf[smallBufSize];
   char fix[bufSize];
-  snprintf(line, bufSize, "<UnknownElement Type=\"%s\" InputChannels=\"%d\" OutputChannels=\"%d\"", 
-           icFixXml(fix, icGetSigStr(buf, bufSize, GetType())), NumInputChannels(), NumOutputChannels());
+  
+  snprintf(line, bufSize*2, "<UnknownElement Type=\"%s\" InputChannels=\"%d\" OutputChannels=\"%d\"",
+           icFixXml(fix, icGetSigStr(buf, smallBufSize, GetType())), NumInputChannels(), NumOutputChannels());
   xml += blanks + line;
 
   if (m_nReserved) {
@@ -164,13 +166,14 @@ icFloatNumber icGetSegPos(const char *str)
 bool CIccFormulaCurveSegmentXml::ToXml(std::string &xml, std::string blanks)
 {
   const size_t bufSize = 256;
-  char buf[bufSize];
+  const size_t smallBufSize = 20;
+  char buf[smallBufSize];
   char line[bufSize];
 
-  snprintf(line, bufSize, "<FormulaSegment Start=\"%s\"", icSegPos(buf, bufSize, m_startPoint));
+  snprintf(line, bufSize, "<FormulaSegment Start=\"%s\"", icSegPos(buf, smallBufSize, m_startPoint));
   xml += blanks + line;
 
-  snprintf(line, bufSize, " End=\"%s\"",icSegPos(buf, bufSize, m_endPoint));
+  snprintf(line, bufSize, " End=\"%s\"", icSegPos(buf, smallBufSize, m_endPoint));
   xml += line;
     
   snprintf(line, bufSize, " FunctionType=\"%d\"", m_nFunctionType);
@@ -273,13 +276,14 @@ public:
 bool CIccSampledCurveSegmentXml::ToXml(std::string &xml, std::string blanks)
 {
   const size_t bufSize = 256;
-  char buf[bufSize];
+  const size_t smallBufSize = 20;
+  char buf[smallBufSize];
   char line[bufSize];
 
-  snprintf(line, bufSize, "<SampledSegment Start=\"%s\"", icSegPos(buf, bufSize, m_startPoint));
+  snprintf(line, bufSize, "<SampledSegment Start=\"%s\"", icSegPos(buf, smallBufSize, m_startPoint));
   xml += blanks + line;
 
-  snprintf(line, bufSize, " End=\"%s\">\n",icSegPos(buf, bufSize, m_endPoint));
+  snprintf(line, bufSize, " End=\"%s\">\n",icSegPos(buf, smallBufSize, m_endPoint));
   xml += line;
 
   CIccFloatArray::DumpArray(xml, blanks+"  ", m_pSamples, m_nCount, icConvertFloat, 8);
@@ -1976,11 +1980,12 @@ bool CIccMpeXmlExtCLUT::ParseXml(xmlNode *pNode, std::string &parseStr)
 bool CIccMpeXmlBAcs::ToXml(std::string &xml, std::string blanks/* = ""*/)
 {
   const size_t bufSize = 256;
-  char line[bufSize];
-  char buf[bufSize], fix[bufSize];
+  char line[bufSize*2];
+  char buf[bufSize/2];
+  char fix[bufSize];
 
-  snprintf(line, bufSize, "<BAcsElement InputChannels=\"%d\" OutputChannels=\"%d\" Signature=\"%s\"", NumInputChannels(), NumOutputChannels(),
-                icFixXml(fix, icGetSigStr(buf, bufSize, m_signature)));
+  snprintf(line, bufSize*2, "<BAcsElement InputChannels=\"%d\" OutputChannels=\"%d\" Signature=\"%s\"", NumInputChannels(), NumOutputChannels(),
+                icFixXml(fix, icGetSigStr(buf, bufSize/2, m_signature)));
   xml += blanks + line;
 
   if (m_nReserved) {
@@ -2031,11 +2036,12 @@ bool CIccMpeXmlBAcs::ParseXml(xmlNode *pNode, std::string &parseStr)
 bool CIccMpeXmlEAcs::ToXml(std::string &xml, std::string blanks/* = ""*/)
 {
   const size_t bufSize = 256;
-  char line[bufSize];
-  char buf[bufSize], fix[bufSize];
+  char line[bufSize*2];
+  char buf[bufSize/2];
+  char fix[bufSize];
 
-  snprintf(line, bufSize, "<EAcsElement InputChannels=\"%d\" OutputChannels=\"%d\" Signature=\"%s\"", NumInputChannels(), NumOutputChannels(),
-    icFixXml(fix, icGetSigStr(buf, bufSize, m_signature)));
+  snprintf(line, bufSize*2, "<EAcsElement InputChannels=\"%d\" OutputChannels=\"%d\" Signature=\"%s\"", NumInputChannels(), NumOutputChannels(),
+    icFixXml(fix, icGetSigStr(buf, bufSize/2, m_signature)));
   xml += blanks + line;
 
   if (m_nReserved) {

